expression/builtin_time.go

```diff
@@ -7113,3 +7114,77 @@ func handleInvalidZeroTime(ctx sessionctx.Context, t types.Time) (bool, error) {
 	}
 	return true, handleInvalidTimeError(ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, t.String()))
 }
+
+// readTSInFunctionClass reads a time window [a, b] and compares it with the latest resolvedTS
+// to determine which TS to use in a read only transaction.
+type readTSInFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *readTSInFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, types.ETDatetime, types.ETDatetime)
+	if err != nil {
+		return nil, err
+	}
+	sig := &builtinReadTSInSig{bf}
+	return sig, nil
+}
+
+type builtinReadTSInSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinReadTSInSig) Clone() builtinFunc {
+	newSig := &builtinTidbParseTsoSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinReadTSInSig) evalInt(row chunk.Row) (int64, bool, error) {
```

Comment:
We want `read_ts_in` as a [**deterministic** function](https://dev.mysql.com/doc/refman/8.0/en/create-procedure.html), which means it always produces the same result for the same input parameters. such as `now()` is deterministic, systime is nondeterministic:
```
mysql> select now(6), now(6), now(6), sysdate(6), sysdate(6), sysdate(6);
+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+
| now(6)                     | now(6)                     | now(6)                     | sysdate(6)                 | sysdate(6)                 | sysdate(6)                 |
+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+
| 2021-04-21 15:44:57.807908 | 2021-04-21 15:44:57.807908 | 2021-04-21 15:44:57.807908 | 2021-04-21 15:44:57.808105 | 2021-04-21 15:44:57.808106 | 2021-04-21 15:44:57.808107 |
+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+----------------------------+
1 row in set (0.00 sec)
```

Reply:
better name `tidb_bound_staleness`